plugin-flex-ts-template-v2/src/feature-library/send-audio-rec-file/custom-components/AudioRecorderPanel/AudioRecorderPanel.tsx

```diff
@@ -0,0 +1,166 @@
+import { StringTemplates } from '../../flex-hooks/strings';
+
+const Mp3Recorder = new MicRecorder({ bitRate: 128 });
+
+export interface OwnProps {
+  showRecorder: boolean;
+  audioFile?: any;
+}
+
+export type Props = OwnProps;
+
+class AudioRecorderPanel extends React.Component<Props> {
+  state = {
+    isRecording: false,
+    audioURL: '',
+    blobURL: '',
+    recorder: null,
+    isBlocked: false,
+    showRecorder: this.props.showRecorder,
+    audioFile: null,
+    confirmSendFile: false,
+    recordingMessage: '',
+    flashing: false,
+  };
+
+  startRec = () => {
+    if (this.state.isBlocked) {
+      console.log('Permission Denied');
+    } else {
+      Mp3Recorder.start()
+        .then(() => {
+          this.setState({ isRecording: true, recordingMessage: 'Recording in progress ' });
+        })
+        .catch((e: any) => console.error(e.message));
+    }
+  };
+
+  stopRec = () => {
+    Mp3Recorder.stop()
+      .getMp3()
+      .then(async ([buffer, blob]: [Uint8Array, Blob]) => {
+        const blobURL = URL.createObjectURL(blob);
+        const conversationSid = Flex.Manager.getInstance().store.getState().flex.chat.messageList.activeConversation;
+
+        // Detach existing file before sending the new one
+        await this.dontSendRec();
+
+        return this.sendRec(buffer, conversationSid).then(() => {
+          this.setState({ blobURL, isRecording: false });
+        });
+      })
+      .catch((e: any) => console.log(e.message));
+  };
+
+  dontSendRec = async () => {
+    try {
+      const { audioFile } = this.state;
+
+      if (audioFile) {
+        const conv = Flex.Manager.getInstance().store.getState().flex.chat.messageList.activeConversation;
+
+        await Actions.invokeAction('DetachFile', {
+          file: audioFile,
+          conversationSid: conv,
+        });
+
+        this.setState({ isRecording: false, audioFile: null });
+      } else {
+        console.log('No audio file to detach.');
+      }
+    } catch (error) {
+      console.log('Error during file detachment:', error);
+    }
+  };
+
+  sendRec = async (buffer: any, conversationSid: any) => {
+    const audioFile = new File(buffer, 'voice-recording.mp3', {
+      type: 'audio/mpeg',
+      lastModified: Date.now(),
+    });
+
+    Actions.invokeAction('AttachFiles', {
+      files: [audioFile],
+      conversationSid,
+    });
+
+    this.setState({ audioFile }, () => {
+      const updatedAudioFile = this.state.audioFile;
+      console.log('Updated audioFile:', updatedAudioFile);
+    });
+  };
+
+  componentDidMount() {
+    if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
+      navigator.mediaDevices
+        .getUserMedia({ audio: true })
+        .then(() => {
+          console.log('Permission granted');
+          this.setState({ isBlocked: false });
+        })
+        .catch(() => {
+          console.log('Permission denied');
+          this.setState({ isBlocked: true });
```

Comment:
If `isBlocked=true`, I think we should display a message in the recorder panel telling the user that they did not grant microphone permission and need to do so before they can record.

Reply:
good Idea... I added a blocked Message in case of Mic access is denied.